### ISSUE TITLE
How to setup error channel for PubSub consumer binding with explicit subscription name? 

### ISSUE THREAD

**Issue by artemptushkin** [GoogleCloudPlatform/spring-cloud-gcp#509]
I have consumer binding defined like:

```yaml
spring.cloud.stream.gcp:
  pubsub:
    bindings:
      ongoing-messages-input:
         consumer:
         subscriptionName: my-pull-transactions
```

I define error handler like this:
```java
    @ServiceActivator(inputChannel = "ongoing-messages-input.errors")
    public void handleError(Message<MessagingException> failedMessage) {
```

The framework creates an error channel out-of-box `my-pull-transactions.errors`, i.e. based on subscription name.

I want it to be `ongoing-messages-input.errors` how to achieve this? It should work as well, all error message from this channel goes to my custom channel.

I **can't rely on the subscription name** as this is a dynamic value for me and depends on the **environment**.

I tried:

* Define 
```yaml
spring.cloud.stream:
  bindings:
    ongoing-messages-input: #channel name
      destination: my-topic-name #topic name
      binder: pubsub
```
* Customize error channel name like this:
```java.
    @Bean
    public ConsumerEndpointCustomizer<PubSubInboundChannelAdapter> messageChannelAdapter() {
        return (endpoint, destinationName, group) -> endpoint.setErrorChannelName("ongoing-messages-input.errors");
    }
```

I get the error:
```
java.lang.IllegalArgumentException: 'beanFactory' must not be null
	at org.springframework.util.Assert.notNull(Assert.java:201)
	at org.springframework.integration.support.channel.ChannelResolverUtils.getChannelResolver(ChannelResolverUtils.java:49)
	at org.springframework.integration.context.IntegrationObjectSupport.getChannelResolver(IntegrationObjectSupport.java:257)
	at org.springframework.integration.endpoint.MessageProducerSupport.getErrorChannel(MessageProducerSupport.java:126)
	at com.google.cloud.spring.pubsub.integration.inbound.PubSubInboundChannelAdapter.getErrorChannel(PubSubInboundChannelAdapter.java:161)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendErrorMessageIfNecessary(MessageProducerSupport.java:252)
	at org.springframework.integration.endpoint.MessageProducerSupport.sendMessage(MessageProducerSupport.java:211)
```

To fix this I tried to define channel bean with a correspondent name - no result:

```java
    @Bean(name = "ongoing-messages-input.errors")
    public MessageChannel errorChannel() {
        return new QueueChannel();
    }
```

Suggestion, it would be **really** helpful if `PubSubInboundChannelAdapter` has a method to return input channel name, as it has reference to this bean. Even fixing this bug, inside `ConsumerEndpointCustomizer` there is no way to properly resolve the new name in case of **multiple** binders. In my example, I have to hardcode, which is not good.

**Workaround:**
Propagate bean factory:

```java
@Bean
    public ConsumerEndpointCustomizer<PubSubInboundChannelAdapter> messageChannelAdapter(ConfigurableApplicationContext applicationContext) {
        return (endpoint, destinationName, group) -> {
            endpoint.setErrorChannelName("ongoing-messages-input.errors");
            endpoint.setBeanFactory(applicationContext.getBeanFactory());
        };
    }
```

I think this workaround should be fixed somewhere in `com.google.cloud.spring.stream.binder.pubsub.PubSubMessageChannelBinder #createConsumerEndpoint` as this class has a reference to app context and a method to get bean factory already.

The suggestion still is open, as it will be a problem for multiple consumer bindings.
I'm open to contribute ✌️

**Comment by dzou** [comment]
Thank you for the report @artemptushkin. On the note about contribution, we would be very happy to consider a PR proposal if you would like to modify something.

Regarding error channel customization, it looks like the naming decision takes place in `PubSubMessageChannelBinder` (as you said) on the call to [registerErrorInfrastructure](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-pubsub-stream-binder/src/main/java/com/google/cloud/spring/stream/binder/pubsub/PubSubMessageChannelBinder.java#L88).

I'm not too familiar with Spring Stream, but I would be open to solutions where you could pass in a custom channel that takes precedence over the default location. It looks a bit tricky tho, studying the code, the error channel base name appears to come from [AbstractMessageChannelBinder#errorsBaseName](https://github.com/spring-cloud/spring-cloud-stream/blob/main/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java#L899) in Spring Stream core. It seems like the naming methods follow some default conventions.

Alternatively, would it be possible to manually create a separate new channel called `ongoing-messages-input.errors` and then forward all the errors from the subscription channels to there?

**Comment by artemptushkin** [comment]
Yes, that's what I've done. This creates a new channel

```
@ServiceActivator(inputChannel = "ongoing-messages-input.errors")
 public void handleError(Message<MessagingException> failedMessage)
```
And I lint it inside `ConsumerEndpointCustomizer<PubSubInboundChannelAdapter>` by hardcoding the channel name. That's is done the wrong way, as `ConsumerEndpointCustomizer` can be used within multiple `PubSubInboundChannelAdapter` but we can't distinguish there which name to use.

Here comes the question from the suggestion. _Can we expose the name of the channel there: something like `pubSubInboundChannelAdapter.getChannelName()` it will allow binding the error by the **channel name** and not the **subscription name**.

**Comment by artemptushkin** [comment]
The second question is, do you consider as a bug that error with `'beanFactory' must not be null`? If yes, I'd fix it.

that I fix with `endpoint.setBeanFactory(applicationContext.getBeanFactory());` on the library client side

**Comment by dzou** [comment]
Ok, I see what you're saying.

- `pubSubInboundChannelAdapter.getChannelName()` seems fine to add, I would be okay with the contribution.
- This `endpoint.setBeanFactory(applicationContext.getBeanFactory());` seems to make sense but will have to double check to make sure it doesn't cause a regression.

Let us check-in with Spring team on how to override error channel.

@artembilan -- Do you have any advice regarding how a user might easily override the default error channel destination for a Spring stream channel binder? We configure a Pub/Sub stream error handler [like this](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-pubsub-stream-binder/src/main/java/com/google/cloud/spring/stream/binder/pubsub/PubSubMessageChannelBinder.java#L88). I'm wondering if there is a standard way in Spring to let users to override this destination or easily forward messages from the default error channel to a custom one.

**Comment by artembilan** [comment]
There is `public MessageChannel getOutputChannel() {`, which you can cast to `NamedComponent` and call its `getBeanName()`.

The `BeanFactory` is indeed a bug: there is a `setApplicationContext(applicationContext);` on the `PubSubInboundChannelAdapter` which has to be called in the `PubSubMessageChannelBinder.createConsumerEndpoint()`  if it is not done yet.

I'm not aware how we can customize the error channel per consumer, but probably that `ConsumerEndpointCustomizer` is the best way to go.
The contract `void configure(E endpoint, String destinationName, String group);` should give you some clues what exactly endpoint we are talking about at the moment of configuration.

**Comment by artemptushkin** [comment]
Good, thanks for the clarification on getting the bean name.

For the history, the proper solution for this case is:
```java
@Bean
public ConsumerEndpointCustomizer<PubSubInboundChannelAdapter> messageChannelAdapter() {
    return (endpoint, destinationName, group) -> {
        var namedComponent = (NamedComponent) endpoint.getOutputChannel();
        String channelName = namedComponent.getBeanName();
        endpoint.setErrorChannelName(channelName + ".errors");
    };
}
```

As for `beanFactory` problem, I see you're open for contributors I take that bug then

**Comment by artemptushkin** [comment]
@dzou Btw, this part of the documentation https://googlecloudplatform.github.io/spring-cloud-gcp/2.0.3/reference/html/index.html#error-handling is not up-to-date, it doesn't compile. The constructor of `PubSubInboundChannelAdapter` expects `PubSubSubscriberOperations` and not that factory, I'd update try to update as well with another PR.

Maybe update the doc with the information about `ConsumerEndpointCustomizer` as well? You use this inside `PubSubBinderConfiguration` so should be helpful

**Comment by dzou** [comment]
> Good, thanks for the clarification on getting the bean name.
> 
> For the history, the proper solution for this case is:
> 
> ```java
> @Bean
> public ConsumerEndpointCustomizer<PubSubInboundChannelAdapter> messageChannelAdapter() {
>     return (endpoint, destinationName, group) -> {
>         var namedComponent = (NamedComponent) endpoint.getOutputChannel();
>         String channelName = namedComponent.getBeanName();
>         endpoint.setErrorChannelName(channelName + ".errors");
>     };
> }
> ```
> 
> As for `beanFactory` problem, I see you're open for contributors I take that bug then

I see, makes sense. Yes we would be happy to accept a contribution for the `beanFactory`! Please feel free to send out a PR.


> @dzou Btw, this part of the documentation https://googlecloudplatform.github.io/spring-cloud-gcp/2.0.3/reference/html/index.html#error-handling is not up-to-date, it doesn't compile. The constructor of `PubSubInboundChannelAdapter` expects `PubSubSubscriberOperations` and not that factory, I'd try to update as well with another PR.
> 
> Maybe update the doc with the information about `ConsumerEndpointCustomizer` as well? You use this inside `PubSubBinderConfiguration` so should be helpful

No problem, I can fix this documentation issue and add a note about `ConsumerEndpointCustomizer`.

**Comment by artemptushkin** [comment]
>No problem, I can fix this documentation issue and add a note about ConsumerEndpointCustomizer.

@dzou good, thanks

**Comment by dzou** [comment]
Doc fix here: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/512

**Comment by dzou** [comment]
Fixed by #515
